presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java

```diff
@@ -311,47 +406,160 @@ private void processPage(Page page)
+                    || (memorySizeBeforeSpill > 0 && memorySize > memorySizeBeforeSpill);
+        }
+
+        private boolean isSpillingEnabled()
+        {
+            return maxEntriesBeforeSpill > 0 || memorySizeBeforeSpill > 0;
+        }
+
+        public Iterator<Page> buildResult()
+        {
+            checkState(!isBusy(), "Previous spill hasn't yet finished");
+
+            if (!spiller.isPresent()) {
+                return buildResultFromMemory();
+            }
+
+            try {
+                // TODO: don't spill here to disk, instead merge disk content with memory content
+                spillToDisk().get();
+                return mergeFromDisk();
+            }
+            catch (InterruptedException | ExecutionException e) {
+                Thread.currentThread().interrupt();
+                throw Throwables.propagate(e);
+            }
+        }
+
+        @Override
+        public void close()
+                throws IOException
+        {
+            closeSpiller();
+        }
+
+        private CompletableFuture<?> spillToDisk()
         {
-            List<Type> types = new ArrayList<>(groupByHash.getTypes());
             for (Aggregator aggregator : aggregators) {
-                types.add(aggregator.getType());
+                aggregator.setOutputPartial();
             }
 
+            if (!spiller.isPresent()) {
+                spiller = Optional.of(spillerFactory.get().create(buildTypes()));
+            }
+
+            // we are starting spilling process with current content of the aggregators and groupByHash...
+            busy = spiller.get().spill(buildResultFromMemory(sortedGroupIds(), buildTypes(), groupByHash, aggregators));
+            // ... and we immediately create new aggregators and groupByHash so we effectively transfer ownership
+            // over groupByHash and aggregators from this thread to spilling thread
+            rebuildAggregators();
+
+            return busy;
+        }
+
+        private Iterator<Page> mergeFromDisk()
+        {
+            checkState(spiller.isPresent());
+
+            convertToMerge();
+            rebuildAggregatorsForMerge();
+
+            MergeSort mergeSort = new MergeSort(groupByTypes, buildIntermediateTypes());
+            Iterator<Page> pages = mergeSort.merge(spiller.get().getSpills());
+
+            return buildResultFromMerge(pages);
+        }
+
+        private Iterator<Page> buildResultFromMerge(Iterator<Page> mergedPages)
+        {
+            return new Iterator<Page>() {
+                private Iterator<Page> resultPages = Collections.emptyIterator();
+
+                @Override
+                public boolean hasNext()
+                {
+                    return mergedPages.hasNext() || resultPages.hasNext();
+                }
+
+                @Override
+                public Page next()
+                {
+                    if (!resultPages.hasNext()) {
+                        rebuildAggregatorsForMerge();
```

Comment:
Constructor is doing exactly the same thing as the simple reset would do. It's not costly, it just drops memory and creates new empty aggregators (empty hash tables). 

Yes it is necessary to do it after every page so that we ensure that memory usage is constant during merge.